<html>
    <div dir = "rtl" style="font-size: 26px; color: blue">
        <b>
            پروژه اول درس هوش مصنوعی
            <br>
            <br>
            Informed and Uninformed Search
        </b>
    </div>
<br>
<br>
<br>
    <div dir = "rtl" style="font-size: 24px; color: gray">
        <b>
            معین کرمی
            <br>
            <br>
            810198540
        </b>
    </div>
</html>

<html>
    <div dir = "rtl" style="font-size: 22px; color: green">
        <b>
            هدف پروژه:
        </b>
    </div>
    <br>
    <br>
    <div dir = "rtl" style = "font-size: 16px">
        <b>
            هدف این پروژه آشنایی با انواع الگوریتم های سرچ
            uninformed 
            مانند 
            BFS, IDS
            و الگوریتم های سرچ 
            informed
            مانند 
            A*, 
            weighted A*
            است. در این پروژه با این الگوریتم ها و مزایا و معایب هرکدام آشنا می‌شویم و می‌بینیم بعضی جواب بهینه و بعضی دیگر جواب نه کاملا بهینه ولی سرعت جواب دهی بیشتر را ارائه می‌دهند.
        </b>
    </div>
</html>

<html>
    <div dir = "rtl" style="font-size: 22px; color: green">
        <b>
            خلاصه ی مسئله:
        </b>
    </div>
    <br>
    <div dir = "rtl" style = "font-size: 16px">
            در این پروژه یک جدول شطرنجی 
            $n \times m$
            داریم که یک مامور در یکی از خانه‌های آن قرار داد. همچنین تعدادی محموله در این جدول وجود دارد که هر محموله باید به خانه ی مشخصی منتقل شود که مامور داخل جدول باید این کار را انجام دهد و خودش در نهایت به خانه مشخصی برود. همچنین این مامور در هر لحظه فقط قادر است یک محموله را انتقال دهد.
            <br>
            در بعضی خانه های این جدول نیز آنتن های وجود دارند که هر یک قدرتی دارند.
            <br>
            مامور ضمن اینکه نمی‌تواند به خانه هایی که شامل آنتن هستند وارد شود، اگر قدرت آنتنی 
            $x$
            باشد، مامور فقط می‌تواند حداکثر 
            $x$
            گام متوالی در خانه هایی بردارد که فاصله ی منهتنی آن ها از آنتن کوچکتر مساوی 
            $x$
            است.
             <br>
            لازم به ذکر است که مامور فقط قادر است به ۴ خانه ی بالا، پایین، راست و چپ خود(در صورت نقض نکردن شرایطی که در بالا گفته شد و بیرون نزدن از جدول)
            حرکت کند.
            <br>
            هدف به دست اوردن حداقل تعداد حرکت لازم است که مامور بتواند محموله ها را به مقصدشان برساند و خود نیز به مقصدش برود.
    </div>
</html>

<html>
    <div dir = "rtl" style="font-size: 22px; color: green">
        <b>
            مدل کردن مسئله:
        </b>
    </div>
    <br>
    <br>
    <div dir = "rtl" style = "font-size: 16px">
            هر استیت در این مسئله شامل این اطلاعات است: مکان مامور، مکان محموله ها، حرکت های انجام شده تا الان و تعداد قدم های متوالی که در حوزه ی هر آنتن برداشته شده است.
            <br>
            <br>
            <b>Initial State:</b>
            این استیت همان استیت شرایط اولیه است که به عنوان ورودی به ما داده میشود.
            <br>
            <br>
            <b>Actions:</b>
            <br>
            ۱) مامور می‌تواند در هر مرحله به خانه ی بالا، چپ، راست، پایین (در صورتی که شرایط مسئله را نقض نکند) حرکت کند.
            <br>
            ۲) اگر مامور در خانه ای باشد که یک محموله نیز در آن خانه است می‌تواند محموله را نیز با خود به یکی از خانه های همسایه منتقل کند.
            <br>
            <br>
            <b>Transition Model:</b>
            <br>
            ۱) اگر مامور در هر جهتی حرکت کند، در نهایت به خانه جدید می‌رود.
            <br>
            ۲) اگر مامور بسته را نیز با خود حمل کند، پس از حرکت در هر جهتی، در نهایت هم خودش و هم بسته به خانه جدید منتقل می‌شوند.
            <br>
            <br>
            <b>Goal State:</b>
            در استیت هدف باید تمام محموله ها و مامور در مقصد خود قرار داشته باشند.
            <br>
            <br>
            <b> Path Cost: </b>
            هر عمل و حرکت مامور، ۱ واحد هزینه دارد.
    </div>
</html>

<html>
    <div dir = "rtl" style="font-size: 20px; color: blue">
        <b>
            import libraries:
        </b>
    </div>
    <br>
</html>

In [1]:
import copy
import collections
import time
import functools
import queue
import heapq

<html>
    <div dir = "rtl" style="font-size: 20px; color: blue">
        <b>
            تعریف کلاس های مورد نیاز:
        </b>
    </div>
    <br>
    <div dir = "rtl" style = "font-size: 16px">
            در ابتدا کلاس ها و لیست های مورد نیاز را تعریف می‌کنیم:
            <br>
            1ـPacket:
            محموله ها در این دسته قرار می‌گیرند.
            <br>
            2_tracker:
            هر آنتن یک 
            tracker 
            است.
    </div>
</html>

In [2]:
trackers = []
packets = []
directions = [[0, 1], [1, 0], [-1, 0], [0, -1]] # all possible move dircections
directions_symbol = ['R', 'D', 'U', 'L']

class Coordinate:
    def __init__(self, listed):
        self.x, self.y = map(int, listed)
        
    def __hash__(self):
        return hash((self.x, self.y))
    
    def __eq__(self, other):
        return (self.x, self.y) == (other.x, other.y)
    
    def __str__(self):
        return f"({self.x}, {self.y})"
    
    def __add__(self, other):
        return Coordinate([self.x + other.x, self.y + other.y])

class Tracker:
    def __init__(self, listed):
        self.coordinate = Coordinate(listed[: 2])
        self.power = int(listed[2])
        self.number = int(listed[3])
    
    def __hash__(self):
        return hash((self.coordinate, self.power, self.number))
    
    def __eq__(self, other):
        return (self.coordinate, self.power, self.number) == (other.coordinate, other.power, other.number)

    def __str__(self):
        return f"Coordinate: {self.coordinate}\npower: {self.power}\nId: {self.number}"

    
class Packet:
    def __init__(self, coordinate, destination, number):
        self.coordinate = coordinate
        self.destination = destination;
        self.number = number

    def __hash__(self):
        return hash((self.coordinate, self.destination, self.number))
    
    def __eq__(self, other):
        return (self.coordinate, self.destination, self.number) == (other.coordinate, other.destination,\
                                                                    other.number)
    
    def __str__(self):
        return f"Id: {self.number}\nCoordinate: {self.coordinate}\ndestination: {self.destination}"

    
class State:
    def __init__(self, agent = Coordinate([0, 0]), remaining_packets = None, steps_in_tracker_zone = None, steps = 0,\
                 carrying_packet_number = None, parent = None, direction = None, number = 0):
        self.agent = agent
        self.remaining_packets = remaining_packets
        self.steps_in_tracker_zone = steps_in_tracker_zone
        self.steps = steps
        self.carrying_packet_number = carrying_packet_number
        self.parent = parent
        self.direction = direction
        self.number = number
    
    def to_string(self):
        string = f"agent: {self.agent}\npackets: "
        string += ' '.join(map(str, self.remaining_packets))
        string += f"\ncarrying packet number: {self.carrying_packet_number}\n"
        string += f"trackers: {self.steps_in_tracker_zone}\n"
        return string
    
    def __eq__(self, other):
        return (self.agent, self.remaining_packets, self.carrying_packet_number,\
                self.steps_in_tracker_zone) == (other.agent, other.remaining_packets,\
                                                other.carrying_packet_number, other.steps_in_tracker_zone)
    
    
    def __hash__(self):
        return hash((self.agent, str(self.carrying_packet_number), sum(self.remaining_packets),\
                     str(self.steps_in_tracker_zone)))
    
    def __str__(self):
        string = "---------\n" + self.to_string()
        string += f"steps:  {str(self.steps)}"
        return string + '\n---------\n'
    
    def __lt__(self, other): #returns true if self.steps is less than other.steps
        return self.steps < other.steps
    
    def fix_carrying_packet(self): #put packet down if we are in its destination and pull it up if 
        #we are in the same place with a packet
        if self.carrying_packet_number != None:
            if self.agent == packets[self.carrying_packet_number].destination:
                self.remaining_packets.remove(self.carrying_packet_number)
                self.carrying_packet_number = None

        if self.carrying_packet_number == None:
            for packet in self.remaining_packets:
                if packets[packet].coordinate == self.agent:
                    self.carrying_packet_number = packet
                    break
    
    def update_steps_in_tracker_zone(self): #returns false if some tracker tracks us (we have stayed in
        #some tracker zone too much)

        if self.steps_in_tracker_zone != None:
            if self.steps_in_tracker_zone[0] == field[self.agent.x, self.agent.y]:
                self.steps_in_tracker_zone[1] += 1
                if self.steps_in_tracker_zone[1] > trackers[self.steps_in_tracker_zone[0]].power:
                    return False
            else:
                self.steps_in_tracker_zone = None

        if self.steps_in_tracker_zone == None and field[self.agent.x, self.agent.y] != None:
            self.steps_in_tracker_zone = [field[self.agent.x, self.agent.y], 1]
        return True

    
    def get_new_states(self): #returns states which can be reached from current state
        new_states = []
        
        for direct, symbol in zip(directions, directions_symbol):
            if self.agent.x + direct[0] >= rows or self.agent.y +\
            direct[1] >= cols or self.agent.x + direct[0] < 0 or self.agent.y\
            + direct[1] < 0:
                continue
   
            new_state = State(remaining_packets =\
                              copy.deepcopy(self.remaining_packets), steps = self.steps,\
                              steps_in_tracker_zone = copy.deepcopy(self.steps_in_tracker_zone),\
                              carrying_packet_number = self.carrying_packet_number)

            new_state.agent = self.agent + Coordinate(direct)
            new_state.steps += 1
            new_state.direction = symbol
            new_state.parent = self.number

            if new_state.update_steps_in_tracker_zone() is False:
                continue
            new_state.fix_carrying_packet()
            new_states.append(new_state) 
        return new_states
    
    def is_goal(self): #returns true if current state is goal state
        if self.agent != agent_destination:
            return False
        if self.remaining_packets != []:
            return False
        return True
    
    def get_T(self): #return T(x)
        return (len(self.remaining_packets) - 1) * min_packet_dest_to_start
    
    def get_h(self): #returns value of heuristic function value
        if len(self.remaining_packets) == 0:
            return get_distance(self.agent, agent_destination)
        min_dist = float('inf')
        sum_dist = 0
        for packet in self.remaining_packets:
            if packet != self.carrying_packet_number:
                sum_dist += get_distance(packets[packet].coordinate, packets[packet].destination)
                min_dist = min(min_dist, get_distance(self.agent, packets[packet].coordinate))
            else:
                sum_dist += get_distance(self.agent, packets[packet].destination)
                min_dist = 0
        return min_dist + sum_dist + self.get_T() + min_dist_to_dest

<html>
    <div dir = "rtl" style="font-size: 20px; color: blue">
        <b>
          تعاریف تابع‌های مورد نیاز:
        </b>
    </div>
    <br>
    <div dir = "rtl" style = "font-size: 16px">
            در اینجا توابعی رو تعریق می‌کنیم که در مراحل مختلف استفاده می‌شوند:
        <br>
        <b>get_distance:</b>
            برای محاسبه فاصله‌ی منهتنی دو خانه.
        <br><br>
    </div>
</html>

In [3]:
def get_distance(loc1, loc2):
    return abs(loc1.x - loc2.x) + abs(loc1.y - loc2.y)

def get_path(cur_state, explored): #returns moves which is done to reach to cur_state
    path = ''
    while cur_state.parent is not None:
        path += cur_state.direction
        cur_state = explored[cur_state.parent]
    return path[::-1]

<html>
    <div dir = "rtl" style="font-size: 20px; color: blue">
        <b>
          خواندن ورودی:
        </b>
    </div>
    <br>
    <div dir = "rtl" style = "font-size: 16px">
            حال باید ورودی ها را خوانده و ذخیره کنیم و استیت اولیه را بسازیم.
    </div>
</html>

In [4]:
file = open("sample_testcases/test_00.txt")
packet_dest_to_start = [] #list of distance of destination and start point of pair of different packets
min_packet_dest_to_start = float('inf') #minimum distance of destination and start point of pair of different packets
min_dist_to_dest = float('inf') #minimum distance between agent destination and packets' destinations

rows, cols = map(int, file.readline().split())

agent = Coordinate(file.readline().split())
agent_destination = Coordinate(file.readline().split())

number_of_trackers, number_of_packets = map(int, file.readline().split())

for i in range(number_of_trackers):
        trackers.append(Tracker(file.readline().split() + [i]))

for i in range(number_of_packets):
        packets.append(Coordinate(file.readline().split()))
for i in range(number_of_packets):
        packets[i] = Packet(packets[i], Coordinate(file.readline().split()), i)

for packet in packets:
    for other_packet in packets:
        if (packet != other_packet):
            packet_dest_to_start.append(get_distance(packet.destination, other_packet.coordinate))
    min_dist_to_dest = min(min_dist_to_dest, get_distance(agent_destination, packet.destination))

packet_dest_to_start.sort()
min_packet_dest_to_start = packet_dest_to_start[0]
    
initial_state = State(agent, [i for i in range(number_of_packets)])

<html>
    <div dir = "rtl" style = "font-size: 16px">
            برای راحتی کار در ادامه،‌مشخص می‌کنیم که هر خانه از جدول داخل منطقه ی کدام آنتن‌ها است. 
    </div>
</html>

In [5]:
field = {}
for i in range(rows):
    for j in range(cols):
        field[i, j] = None
        for tracker in trackers:
            if get_distance(Coordinate([i, j]), tracker.coordinate) <= tracker.power:
                field[i, j] = tracker.number

<html>
    <div dir = "rtl" style="font-size: 22px; color: green">
        <b>
           BFS:
        </b>
    </div>
    <br>
    <div dir = "rtl" style = "font-size: 16px">
            این الگوریتم یک الگوریتم
        uninformed
        است که راس های مختلف را بر اساس ارتفاع آن‌ها می‌بیند و وقتی تمام راس های  موجود در یک ارتفاع را دید، به ارتفاع بعدی رفته و رئوس ارتفاع جدید را می‌بیند. در این الگوریتم همواره استیت ها و راس های مرزی را در یک صف نگه داشته و به ترتیب از سر صف گسترش می‌دهیم. همچنین برای جلوگیری از دیدن راس های تکراری، راس های گسترش داده شده را نیز در یک 
        set
        نگه می‌داریم.
        <br>
        مزایا: این الگوریتم کامل است و جواب بهینه را ارائه می‌دهد  و نسبت به 
        IDS
        سریع تر است.
        <br>
        معایب: از آنجایی که این الگوریتم
        uninformed
        است، به حافظه و زمان زیادی نیاز دارد و از 
        A*
        کند تر است.
        <br>
        Time complexity: O(b^d)
        <br>
        Space complexity: O(b^d)
    </div>
</html>

In [6]:
def bfs(initial_state):
    explored = set()
    frontier = queue.Queue()
    seen_states = 0
    explored_dict = {}
    
    frontier.put(copy.deepcopy(initial_state))

    while frontier.empty() == False:
        cur_state = frontier.get()
        cur_state.number = seen_states
        
        if cur_state in explored:
            continue

        seen_states += 1
        
        if cur_state.is_goal():
            return (cur_state, seen_states, get_path(cur_state, explored_dict))

        explored.add(cur_state)
        explored_dict[cur_state.number] = cur_state

        new_states = cur_state.get_new_states()

        for state in new_states:
            if state not in explored:
                frontier.put(state)

    return (None, seen_states, None)

In [7]:
start_time = time.time()
(goal_state, seen_states, path) = bfs(initial_state)
end_time = time.time()

print("Executation Time:", end_time - start_time, "\nSeen States:", seen_states)
# print(goal_state)
print("path: ", path)
print("steps: ", len(path))

Executation Time: 0.36055421829223633 
Seen States: 6947
path:  RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR
steps:  48


<table style = "width: 85%; font-size: 14px; margin-left:auto; margin-right:auto;">
    <tr style = "text-align: left">
        <th style = "text-align: left">Test</th>
        <th style = "text-align: left">Goal Depth</th>
        <th style = "text-align: left">Seen States</th>
        <th style = "text-align: left">Mean Execution Time</th>
        <th style = "text-align: left">Path</th>
    </tr>
    <tr>
        <td style = "text-align: left">1</td>
        <td style = "text-align: left">48</td>
        <td style = "text-align: left">6947</td>
        <td style = "text-align: left"> 0.339</td>
        <td style = "text-align: left"> RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR</td>
    </tr>
    <tr>
        <td style = "text-align: left">2</td>
        <td style = "text-align: left">52</td>
        <td style = "text-align: left">1391</td>
        <td style = "text-align: left"> 0.083</td>
        <td style = "text-align: left"> RRRRDDDDDDDDLLLRRRUUURURUDLDLDDDLRRRRRUUUUDDDDLLLRRR</td>
    </tr>
    <tr>
        <td style = "text-align: left">3</td>
        <td style = "text-align: left">34</td>
        <td style = "text-align: left">392</td>
        <td style = "text-align: left"> 0.024</td>
        <td style = "text-align: left"> RRRRRRRDDDDLLLLDLLLRDDRUURURRRDRDD</td>
    </tr>
        <tr>
        <td style = "text-align: left">4</td>
        <td style = "text-align: left">66</td>
        <td style = "text-align: left">3976</td>
        <td style = "text-align: left"> 0.210</td>
            <td style = "text-align: left"> RRDDDDDDDDLLRDRRRRRRRRUUUUUUUUULLDDDDDLLDDDDLLLLRRRRUUUUURRRRDDDDD</td>
    </tr>
</table>

<html>
    <div dir = "rtl" style="font-size: 22px; color: green">
        <b>
           IDS:
        </b>
    </div>
    <br>
    <div dir = "rtl" style = "font-size: 16px">
            این الگوریتم یک الگوریتم
        uninformed
        است و مانند الگوریتم
        BFS
        کامل و بهینه است ولی نسبت به آن به زمان بیشتری برای محاسبات نیاز دارد اما به حافظه‌ی کمتری نیاز دارد، چون تشابه زیادی به 
       DFS
        دارد و تنها تفاوت آن در این است که هر بار مشخص است تا چه عمقی پیش می‌رویم.
        <br>
        در پیاده سازی زیر تابع 
        ids
        هر بار یک عمق را به عنوان حداکثر عمق پیمایش مشخص می‌کند و این مقدار را به تابع
        dfs
        می‌دهد تا 
        dfs
        روی درخت تا این ارتفاع مشخص شده انجام شود.
        <br>
        مزایا: کامل و اپتیمال است و به فضای کمی نسبت به دو الگوریتم دیگر نیاز دارد.
        <br>
        معایب: نسبت به بقیه الگوریتم ها کند تر است.
        <br>
        Time complexity: O(b^d)
        <br>
        Space complexity: O(b^d)
        <br>
        توجه کنید با اینکه به صورت تئوری پیچیدگی زمانی و حافظه این الگوریتم با 
        BFS
        برابر است ولی در عمل این الگوریتم حافظه ی کمتر و زمان بیشتری می‌خواهد.
    </div>
</html>

In [8]:
def dfs(cur_state, explored, seen_states, explored_dict, max_depth):
    if (cur_state, cur_state.steps) in explored or cur_state.steps > max_depth:
        return None
    cur_state.number = seen_states[0]
    explored_dict[seen_states[0]] = cur_state
    explored.add((cur_state, cur_state.steps))
    seen_states[0] += 1
    if cur_state.is_goal():
        return cur_state
    
    new_states = cur_state.get_new_states()
    
    for state in new_states:
        if (state, state.steps) not in explored:
            goal_state = dfs(state, explored, seen_states, explored_dict, max_depth)
            if goal_state is not None:
                return goal_state
    return None

def ids(initial_state):
    depth = 0
    seen_states = [0]
    while True:
        explored = set()
        explored_dict = {}
        goal_state = dfs(initial_state, explored, seen_states, explored_dict, depth)
        if goal_state is not None:
            break
        depth += 1;
    
    return (goal_state, seen_states, get_path(goal_state, explored_dict))

In [9]:
start_time = time.time()
(goal_state, seen_states, path) = ids(initial_state)
finish_time = time.time()

print("Executation Time:", finish_time - start_time, "\nSeen States:", seen_states[0])
# print(goal_state)
print("path: ", path)
print("steps: ", len(path))

Executation Time: 5.684725046157837 
Seen States: 132423
path:  RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR
steps:  48


<table style = "width: 85%; font-size: 14px; margin-left:auto; margin-right:auto;">
    <tr style = "text-align: left">
        <th style = "text-align: left">Test</th>
        <th style = "text-align: left">Goal Depth</th>
        <th style = "text-align: left">Seen States</th>
        <th style = "text-align: left">Mean Execution Time</th>
        <th style = "text-align: left">Path</th>
    </tr>
    <tr>
        <td style = "text-align: left">1</td>
        <td style = "text-align: left">48</td>
        <td style = "text-align: left">132423</td>
        <td style = "text-align: left">5.19</td>
        <td style = "text-align: left"> RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR</td>
    </tr>
    <tr>
        <td style = "text-align: left">2</td>
        <td style = "text-align: left">52</td>
        <td style = "text-align: left">247031</td>
        <td style = "text-align: left"> 9.69</td>
        <td style = "text-align: left"> RRRRDDDDDDDDLLLRRRUUURURUDLDLDDDLRRRRRUUUUDDDDLLLRRR</td>
    </tr>
    <tr>
        <td style = "text-align: left">3</td>
        <td style = "text-align: left">34</td>
        <td style = "text-align: left">26010</td>
        <td style = "text-align: left"> 1.00</td>
        <td style = "text-align: left"> RRRRRRRDDDDLLLLDLLLRDDRUURURRRDRDD</td>
    </tr>
        <tr>
        <td style = "text-align: left">4</td>
        <td style = "text-align: left">66</td>
        <td style = "text-align: left">964088</td>
        <td style = "text-align: left"> 42.44</td>
            <td style = "text-align: left"> RRDDDDDDDDLLRDRRRRRRRRUUUUUUUUULLDDDDDLLDDDDLLLLRRRRUUUUURRRRDDDDD</td>
    </tr>
</table>

<html>
    <div dir = "rtl" style="font-size: 22px; color: green">
        <b>
           *A:
        </b>
    </div>
    <br>
    <div dir = "rtl" style = "font-size: 16px">
        این الگوریتم بر خلاف دو الگوریتم قبلی یک الگوریتم 
        informed
        است.
        <br>
        این الگوریتم در واقع ترکیبی از دو الگوریتم 
        BFS, Dijkstra
        است. بدین ترتیب در هر مرحله راسی گسترش داده می‌شود که تا کنون هزینه ی زیادی نبرده باشد.
        <br>
        از طرفی دیگر در این الگوریتم تابعی به نام
        heuristic
        تعریف می‌شود که به هر راس عددی نسبت می دهد که درواقع پیش بینی هزینه لازم برای رسیدن از آن راس به راس هدف است.
        این تابع با 
        $h(n)$
        نمایش داده می‌شود.
        <br>
        همچنین فرض کنید هزینه لازم برای رسیدن به راس 
        n
        برابر 
        $g(n)$
        باشد. در این صورت تعریف می‌کنیم:
        $$f(n) = h(n) + g(n)$$
        در هر مرحله ی این الگوریتم راسی را انتخاب می‌کنیم و گسترش می‌دهیم که حداقل
        $f(n)$
        را داشته باشد.
        <br>
        این الگوریتم کامل است.
        ولی با توجه به 
        $h(n)$
        تعریف شده می‌تواند بهینه باشد یا نباشد.
        به بیان دقیق تر این الگوریتم زمانی بهینه و اپتیمال است که
        $h(n)$
        ،
        Addmissible
        باشید یعنی هزینه پیش بینی شده برای هر راس تا رسیدن به راس هدف کوچکتر مساوی هزینه واقعی باشد.
        <br>
        اگر 
        $h(n)$
        ، Consistent 
        باشد این الگوریتم 
        optimally efficient
        است، یعنی حداقل راس مورد نیاز را بست می‌ده
        <br>
        مزایا: 
        <br>
        کامل است
        <br>
        با تعریف کردن 
        $h(n)$
        به صورت 
        Addmissible
        این تابع اپتیمال است
        <br>
        در عمل نسبت به دو الگوریتم پیشین سریع تر است.
        <br>
        معایب: در صورتی که تابع 
        heuristic
        به خوبی تعریف نشود، این الگوریتم می‌تواند مانند 
        BFS
        کند باشد و یا اپتیمال نباشد.
        <br>
        Time Complexity: تعداد رئوس گسترش داده شده
        <br>
        Space Complexity: نمایی
        <br><br>
        در پیاده سازی زیر از یک 
        heap
        برای انتخاب راس با حداقل
        $f(n)$
        در هر مرحله استفاده میشود.
    </div>
</html>

<html>
    <div dir = "rtl" style="font-size: 22px; color: green">
        <b>
           Heuristic Declaration:
        </b>
    </div>
    <br>
    <div dir = "rtl" style = "font-size: 16px">
        تابع 
        Heuristic
        برای یک استیت را به این صورت برای دو حالت تعریف می‌کنیم:
        <br>
        حالت اول: حداقل یک محموله به مقصد خود نرسیده باشد، تعداد محموله هایی که به مقصدشان نرسیده‌اند را
        $x > 0$
        در نظر بگیرید.:
        <br>
        در این حالت تابع
        Heuristic
        برابر مجموع فاصله ی منهتنی مکان فعلی محموله‌ها از مقصدشان به اضافه فاصله ی منهتنی مامور از نزدیک ترین محموله ای که به مقصدش نرسیده است به اضافه ی 
        $T(x)$
        به اضافه فاصله منهتنی نزدیک ترین مقصد محموله ها به مقصد مامور تا مقصد مامور است.
        <br>
        تعریف
        $T(x)$
        :
        یک لیست می‌سازیم که به ازای هر دو محموله متفاوت، شامل فاصله ی منهتنی مقصد اولی از مبدا دومی و مبدا اولی از مقصد دومی است.
        <br>
        $T(x)$
        برابر 
        $x-1$
        ضربدر کوچک ترین عضو این لیست است.
        <br>
        <br>
        حالت دوم: تمام محموله ها به مقصدشان رسیده باشند:
        <br>
        در این صورت تابع 
        Heristic
        را برابر فاصله‌ی منهتنی مامور از مقصدش تعریف می‌کنیم.
        <br>
        <br>
        این تابع
        Addmissible 
        است. اثبات:
        <br>
        در حالتی که حداقل یک محموله به مقصد نرسیده باشد چون
        <br>
        (۱)مامور می‌تواند در هر لحظه فقط یک جسم را جا به جا کند و
        <br>
        (۲)برای جا به جایی یک محموله باید به مکان آن برود و 
        <br>
        (۳)پس از رساندن یک محموله به مقصد در صورت وجود محموله ای که در مقصدش نیست باید به مبدا آن محموله برود و 
        <br>
        (۴)در نهایت نیز باید از مقصد آخرین محموله به مقصد خودش برود پس حداقل به مجموع 
        <br>(۱)فاصله ی منهتنی مکان فعلی  محموله‌ها از مقصدشان به اضافه 
        <br>
        (۲)فاصله ی منهتنی مامور از نزدیک ترین  محموله ای که به مقصدش نرسیده است 
        <br>(۳)به اضافه 
        $x-1$
        برابر کوچکترین فاصله ی میان مبدا و مقصد محموله های مختلف 
        <br>
        (۴)به اضافه ی فاصله ی منهتنی نزدیک ترین مقصد محموله به مقصد مامور تا مقصد مامور، 
        <br>
        گام نیاز داریم.
        <br>
        <br>
        در حالت دوم چون مامور باید به مقصدش برسد و فقط می‌تواند در ۴ جهت حرکت کند پس حداقل به فاصله ی منهتنی اش با مقدصش گام نیاز دارد تا به مقصدش برسد.
        <br>
        <br>
        این تابع
        Consistent
        است.
        اثبات:
        <br>
        با توجه به تعریف تابع درمیابیم برای کاهش یک واحد از مقدار تابع باید حداقل یک گام انجام دهیم.
        در نتیجه اگر از یک استیت مانند 
        A
        به 
        B
        با 
        n
        حرکت برویم داریم:
        $$Cost(A to B) = n >= h(A) - h(B)$$
        <br>
        اثبات اینکه در هر گام حداکثر یک واحد از تابع کاهش می‌یابد:
        در اینجا دو حالت داریم:
        <br>
        ۱: اگر هیچ محموله ای باقی نمانده باشد مامور در هرگام حداکثر می‌تواند یک واحد به مقصد خود نزدیک شود، پس با توجه به تعریف تابع، در هر مرحله حداکثر یک واحد از مقدار تابع کم می‌شود.
        <br>
        ۲: حالتی که در آن حداقل یک محموله مانده باشد:
        این حالت را نیز به دو حالت تقسیم می‌کنیم:
        <br>
        ۲-۱: اگر قبل از انجام یک گام مامور در حال جا به جایی هیچ محموله ای نباشد بعد از انجام این گام تنها مقدار از تابع که می‌تواند تغییر کند فاصله ی مامور از نزدیک ترین محموله به خود است که این مقدار نیز حداکثر یک واحد تغییر می‌کند.
        <br>
        ۲-۲: حالتی که مامور در حال انتقال یک محموله باشد را نیز دو حالت می‌کنیم:
        <br>
        ۲-۲-۱: اگر مامور در حال انتقال یک محموله باشد و بعد از انجام یک حرکت به مقصد محموله نرسد، تنها مقدار فاصله ی محموله ها از مقصدشان در تابع و حداکثر به اندازه ی یک واحد کاهش پیدا می‌کند.
        <br>
        ۲-۲-۲: اگر محموله بعد از یک گام به مقصد خود برسد در اینجا نیز دو حالت داریم:
        <br>
        ۲-۲-۲-۱ : اگر این محموله آخرین محموله باشد قبل از انجام این گام مقدار تابع برابر ۱ + فاصله ی مقصد مامور از نزدیک ترین مقصد محموله ها به مقصد مامور  است و بعد از این گام مقدار تابع برابر فاصله ی مامور تا مقصد خود است که حداقل برابر فاصله ی مقصد مامور از نزدیک ترین مقصد محموله ها به مقصد مامور  است. پس در اینجا نیز مقدار تابع حداکثر یک واحد کاهش می‌یابد.
        <br>
        ۲-۲-۲-۲: حال اگر این محموله آخرین محموله نباشد سه مقدار از تابع تغییر می‌کند. اولی مقدار 
        $T(x)$
        است که به 
        $T(x-1)$
        کاهش میابد
         و دومی فاصله‌ی مامور از نزدیک ترین محموله به خود است که حداقل به اندازه‌ی
        $T(x) - T(x - 1)$
        افزایش می‌یابد(توجه کنید این مقدار برابر با حداقل فاصله میان مبدا و مقصد دو محموله ی مختلف است و 
        $x$
        برابر تعداد محموله های باقی مانده است.)
        <br>
        سومین مقداری که تغییر می‌کند فاصله ی محموله های باقی مانده از مقصدشان است که یک واحد کاهش می‌یابد.
        <br>
        پس در این حالت هم مقدار تابع حداکثر یک واحد کاهش می‌یابد و حکم اثبات می‌شود.
        <br>
    </div>
</html>

In [10]:
def A_star(initial_state, weight = 1):
    explored = set()
    frontier = [(initial_state.get_h() * weight, initial_state)]
    seen_states = 0
    explored_dict = {}
    front_set = set()
    
    while len(frontier):
        cur_state = heapq.heappop(frontier)[1]
        cur_state.number = seen_states
        
        if cur_state in explored:
            continue

        seen_states += 1
        
        if cur_state.is_goal():
            return (cur_state, seen_states, get_path(cur_state, explored_dict))

        explored.add(cur_state)
        explored_dict[cur_state.number] = cur_state
        
        new_states = cur_state.get_new_states()
        
        for state in new_states:
            if state not in explored and state not in front_set:
                heapq.heappush(frontier, (state.steps + state.get_h() * weight, state))
                front_set.add(state)

    return (None, seen_states, None)

In [11]:
start_time = time.time()
(goal_state, seen_states, path) = A_star(initial_state)
end_time = time.time()

print("Executation Time:", end_time - start_time, "\nSeen States:", seen_states)
# print(goal_state)
print("path: ", path)
print("steps: ", len(path))

Executation Time: 0.005419015884399414 
Seen States: 53
path:  RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR
steps:  48


<table style = "width: 85%; font-size: 14px; margin-left:auto; margin-right:auto;">
    <tr style = "text-align: left">
        <th style = "text-align: left">Test</th>
        <th style = "text-align: left">Goal Depth</th>
        <th style = "text-align: left">Seen States</th>
        <th style = "text-align: left">Mean Execution Time</th>
        <th style = "text-align: left">Path</th>
    </tr>
    <tr>
        <td style = "text-align: left">1</td>
        <td style = "text-align: left">48</td>
        <td style = "text-align: left">53</td>
        <td style = "text-align: left">0.0059</td>
        <td style = "text-align: left"> RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR</td>
    </tr>
    <tr>
        <td style = "text-align: left">2</td>
        <td style = "text-align: left">52</td>
        <td style = "text-align: left">532</td>
        <td style = "text-align: left"> 0.042</td>
        <td style = "text-align: left"> RRDDRRDDDDDDLLLRRRUUUURRUDLDLDLDDRRRRURUUUDDDDLLLRRR</td>
    </tr>
    <tr>
        <td style = "text-align: left">3</td>
        <td style = "text-align: left">34</td>
        <td style = "text-align: left">106</td>
        <td style = "text-align: left"> 0.009</td>
        <td style = "text-align: left"> RRRRRRRDLLLDLDDDLLLRDDRUURURRRDRDD</td>
    </tr>
        <tr>
        <td style = "text-align: left">4</td>
        <td style = "text-align: left">66</td>
        <td style = "text-align: left">990</td>
        <td style = "text-align: left"> 0.073</td>
            <td style = "text-align: left"> RRDDDDDDLLDDDRRRRUUUURURUUURURRLLDDDDLLDDLDDLDLLRRRRUUUUURRRRDDDDD</td>
    </tr>
</table>

<html>
    <div dir = "rtl" style="font-size: 22px; color: green">
        <b>
           Weighted *A:
        </b>
    </div>
    <br>
    <div dir = "rtl" style = "font-size: 16px">
            این الگوریتم دقیقا مانند الگوریتم 
        *A
        است و تنها تفاوت آن این است که به تابع 
        Heuristic
        یک ضریب 
        $\alpha > 1$
        اختصاص می‌دهیم.
        و داریم:
        $$f(n) = g(n) + h(n) \times \alpha$$
        <br>
        مزایا:
        این الگوریتم نسبت به 
        الگوریتم های قبلی
        سریع تر عمل می‌کند.
        <br>
        معایب:
        ممکن است این الگوریتم اپتیمال نباشد پاسخ بهینه را به دست نیاورد. و در عوض در زمانی سریع تر پاسخی نزدیک به پاسخ بهینه به دست آورد.
        <br><br>
        در واقع در این الگوریتم یک تریدآف میان زمان اجرا و بهینه بودن جواب داریم.
        <br>
        Time Complexity: تعداد رئوس گسترش داده شده
        <br>
        Space Complexity: نمایی
        <br><br><br><br>
        برای پیاده سازی این الگوریتم یک ضریب به 
        heuristic
        اضافه می‌کنیم و الگوریتم 
        *A
        را اجرا می‌کنیم.
    </div>
</html>

<html>
    <div dir = "rtl" style="font-size: 22px; color: green">
        <b>
           $\alpha = 2$
        </b>
    </div>
</html>

In [12]:
start_time = time.time()
(goal_state, seen_states, path) = A_star(initial_state, 2)
end_time = time.time()

print("Executation Time:", end_time - start_time, "\nSeen States:", seen_states)
# print(goal_state)
print("path: ", path)
print("steps: ", len(path))

Executation Time: 0.004057407379150391 
Seen States: 49
path:  RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR
steps:  48


<table style = "width: 85%; font-size: 14px; margin-left:auto; margin-right:auto;">
    <tr style = "text-align: left">
        <th style = "text-align: left">Test</th>
        <th style = "text-align: left">Goal Depth</th>
        <th style = "text-align: left">Seen States</th>
        <th style = "text-align: left">Mean Execution Time</th>
        <th style = "text-align: left">Path</th>
    </tr>
    <tr>
        <td style = "text-align: left">1</td>
        <td style = "text-align: left">48</td>
        <td style = "text-align: left">49</td>
        <td style = "text-align: left">0.0052</td>
        <td style = "text-align: left"> RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR</td>
    </tr>
    <tr>
        <td style = "text-align: left">2</td>
        <td style = "text-align: left">52</td>
        <td style = "text-align: left">93</td>
        <td style = "text-align: left"> 0.0061</td>
        <td style = "text-align: left"> RRRRDDDDDDDDLLLRRRUUURURUDLDLDDDLRRRRRUUUUDDDDLLLRRR</td>
    </tr>
    <tr>
        <td style = "text-align: left">3</td>
        <td style = "text-align: left">34</td>
        <td style = "text-align: left">46</td>
        <td style = "text-align: left"> 0.0024</td>
        <td style = "text-align: left"> RRRRRRRDDDLLLDLDLLLRDDRUURURRRDRDD</td>
    </tr>
        <tr>
        <td style = "text-align: left">4</td>
        <td style = "text-align: left">66</td>
        <td style = "text-align: left">158</td>
        <td style = "text-align: left"> 0.0085</td>
            <td style = "text-align: left"> RRDDDDDDDDLLRDRRRRRRRRUUUUUUUUULLDDDDDLLDDDDLLLLRRRRUUUUURRRRDDDDD</td>
    </tr>
</table>

<html>
    <div dir = "rtl" style="font-size: 22px; color: green">
        <b>
           $\alpha = 10$
        </b>
    </div>
</html>

In [13]:
start_time = time.time()
(goal_state, seen_states, path) = A_star(initial_state, 10)
end_time = time.time()

print("Executation Time:", end_time - start_time, "\nSeen States:", seen_states)
# print(goal_state)
print("path: ", path)
print("steps: ", len(path))

Executation Time: 0.003958463668823242 
Seen States: 49
path:  RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR
steps:  48


<table style = "width: 85%; font-size: 14px; margin-left:auto; margin-right:auto;">
    <tr style = "text-align: left">
        <th style = "text-align: left">Test</th>
        <th style = "text-align: left">Goal Depth</th>
        <th style = "text-align: left">Seen States</th>
        <th style = "text-align: left">Mean Execution Time</th>
        <th style = "text-align: left">Path</th>
    </tr>
    <tr>
        <td style = "text-align: left">1</td>
        <td style = "text-align: left">48</td>
        <td style = "text-align: left">49</td>
        <td style = "text-align: left">0.0060</td>
        <td style = "text-align: left"> RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR</td>
    </tr>
    <tr>
        <td style = "text-align: left">2</td>
        <td style = "text-align: left">54</td>
        <td style = "text-align: left">79</td>
        <td style = "text-align: left"> 0.0120</td>
        <td style = "text-align: left"> RRRRDDDDDDDDLLLRRRRRRRUUUUDDDDLLLULUURURUDLDLDDDLRRRRR</td>
    </tr>
    <tr>
        <td style = "text-align: left">3</td>
        <td style = "text-align: left">36</td>
        <td style = "text-align: left">42</td>
        <td style = "text-align: left"> 0.0039</td>
        <td style = "text-align: left"> RRRRRRRDDDDDLLULLDLLLRDDRUURURRRDRDD</td>
    </tr>
        <tr>
        <td style = "text-align: left">4</td>
        <td style = "text-align: left">66</td>
        <td style = "text-align: left">119</td>
        <td style = "text-align: left"> 0.0067</td>
            <td style = "text-align: left"> RRDDDDDDDDLLRDRRRRRRRRUUUUUUUUULLDDDDDLLDDDDLLLLRRRRUUUUURRRRDDDDD</td>
    </tr>
</table>

<html>
    <div style="font-size: 22px; color: green">
        <b>
           $TEST 00$
        </b>
    </div>
</html>

<table style = "width: 85%; font-size: 14px; margin-left:auto; margin-right:auto;">
    <tr style = "text-align: left">
        <th style = "text-align: left"></th>
        <th style = "text-align: left">طول جواب</th>
        <th style = "text-align: left">تعداد استیت های دیده شده</th>
        <th style = "text-align: left">میانگین زمان اجرا</th>
        <th style = "text-align: left">مسیر پیدا شده توسط الگوریتم</th>
    </tr>
    <tr>
        <td style = "text-align: left">BFS</td>
        <td style = "text-align: left">48</td>
        <td style = "text-align: left">6947</td>
        <td style = "text-align: left">0.339</td>
        <td style = "text-align: left">RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR</td>
    </tr>
    <tr>
        <td style = "text-align: left">IDS</td>
        <td style = "text-align: left">48</td>
        <td style = "text-align: left">132423</td>
        <td style = "text-align: left"> 5.19</td>
        <td style = "text-align: left"> RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR</td>
    </tr>
    <tr>
        <td style = "text-align: left">A*</td>
        <td style = "text-align: left">48</td>
        <td style = "text-align: left">53</td>
        <td style = "text-align: left"> 0.0059</td>
        <td style = "text-align: left"> RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR</td>
    </tr>
        <tr>
        <td style = "text-align: left">A* $\alpha = 2$</td>
        <td style = "text-align: left">48</td>
        <td style = "text-align: left">49</td>
        <td style = "text-align: left"> 0.0052</td>
            <td style = "text-align: left"> RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR</td>
    </tr>
            <tr>
        <td style = "text-align: left">A* $\alpha = 10$</td>
        <td style = "text-align: left">48</td>
        <td style = "text-align: left">49</td>
        <td style = "text-align: left"> 0.0060</td>
            <td style = "text-align: left"> RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR</td>
    </tr>
</table>

<html>
    <div style="font-size: 22px; color: green">
        <b>
           $TEST 01$
        </b>
    </div>
</html>

<table style = "width: 85%; font-size: 14px; margin-left:auto; margin-right:auto;">
    <tr style = "text-align: left">
        <th style = "text-align: left"></th>
        <th style = "text-align: left">طول جواب</th>
        <th style = "text-align: left">تعداد استیت های دیده شده</th>
        <th style = "text-align: left">میانگین زمان اجرا</th>
        <th style = "text-align: left">مسیر پیدا شده توسط الگوریتم</th>
    </tr>
    <tr>
        <td style = "text-align: left">BFS</td>
        <td style = "text-align: left">52</td>
        <td style = "text-align: left">1391</td>
        <td style = "text-align: left">0.083</td>
        <td style = "text-align: left">RRRRDDDDDDDDLLLRRRUUURURUDLDLDDDLRRRRRUUUUDDDDLLLRRR</td>
    </tr>
    <tr>
        <td style = "text-align: left">IDS</td>
        <td style = "text-align: left">52</td>
        <td style = "text-align: left">247031</td>
        <td style = "text-align: left"> 9.69</td>
        <td style = "text-align: left"> RRRRDDDDDDDDLLLRRRUUURURUDLDLDDDLRRRRRUUUUDDDDLLLRRR</td>
    </tr>
    <tr>
        <td style = "text-align: left">A*</td>
        <td style = "text-align: left">52</td>
        <td style = "text-align: left">532</td>
        <td style = "text-align: left"> 0.042</td>
        <td style = "text-align: left"> RRDDRRDDDDDDLLLRRRUUUURRUDLDLDLDDRRRRURUUUDDDDLLLRRR</td>
    </tr>
        <tr>
        <td style = "text-align: left">A* $\alpha = 2$</td>
        <td style = "text-align: left">52</td>
        <td style = "text-align: left">93</td>
        <td style = "text-align: left"> 0.0061</td>
            <td style = "text-align: left"> RRRRDDDDDDDDLLLRRRUUURURUDLDLDDDLRRRRRUUUUDDDDLLLRRR</td>
    </tr>
            <tr>
        <td style = "text-align: left">A* $\alpha = 10$</td>
        <td style = "text-align: left">54</td>
        <td style = "text-align: left">79</td>
        <td style = "text-align: left"> 0.0120</td>
            <td style = "text-align: left"> RRRRDDDDDDDDLLLRRRRRRRUUUUDDDDLLLULUURURUDLDLDDDLRRRRR</td>
    </tr>
</table>

<html>
    <div style="font-size: 22px; color: green">
        <b>
           $TEST 02$
        </b>
    </div>
</html>

<table style = "width: 85%; font-size: 14px; margin-left:auto; margin-right:auto;">
    <tr style = "text-align: left">
        <th style = "text-align: left"></th>
        <th style = "text-align: left">طول جواب</th>
        <th style = "text-align: left">تعداد استیت های دیده شده</th>
        <th style = "text-align: left">میانگین زمان اجرا</th>
        <th style = "text-align: left">مسیر پیدا شده توسط الگوریتم</th>
    </tr>
    <tr>
        <td style = "text-align: left">BFS</td>
        <td style = "text-align: left">34</td>
        <td style = "text-align: left">392</td>
        <td style = "text-align: left">0.024</td>
        <td style = "text-align: left">RRRRRRRDDDDLLLLDLLLRDDRUURURRRDRDD</td>
    </tr>
    <tr>
        <td style = "text-align: left">IDS</td>
        <td style = "text-align: left">34</td>
        <td style = "text-align: left">26010</td>
        <td style = "text-align: left"> 1.00</td>
        <td style = "text-align: left"> RRRRRRRDDDDLLLLDLLLRDDRUURURRRDRDD</td>
    </tr>
    <tr>
        <td style = "text-align: left">A*</td>
        <td style = "text-align: left">34</td>
        <td style = "text-align: left">106</td>
        <td style = "text-align: left"> 0.009</td>
        <td style = "text-align: left"> RRRRRRRDLLLDLDDDLLLRDDRUURURRRDRDD</td>
    </tr>
        <tr>
        <td style = "text-align: left">A* $\alpha = 2$</td>
        <td style = "text-align: left">34</td>
        <td style = "text-align: left">46</td>
        <td style = "text-align: left"> 0.0024</td>
            <td style = "text-align: left"> RRRRRRRDDDLLLDLDLLLRDDRUURURRRDRDD</td>
    </tr>
            <tr>
        <td style = "text-align: left">A* $\alpha = 10$</td>
        <td style = "text-align: left">36</td>
        <td style = "text-align: left">42</td>
        <td style = "text-align: left"> 0.0039</td>
            <td style = "text-align: left"> RRRRRRRDDDDDLLULLDLLLRDDRUURURRRDRDD</td>
    </tr>
</table>

<html>
    <div style="font-size: 22px; color: green">
        <b>
           $TEST 03$
        </b>
    </div>
</html>

<table style = "width: 85%; font-size: 14px; margin-left:auto; margin-right:auto;">
    <tr style = "text-align: left">
        <th style = "text-align: left"></th>
        <th style = "text-align: left">طول جواب</th>
        <th style = "text-align: left">تعداد استیت های دیده شده</th>
        <th style = "text-align: left">میانگین زمان اجرا</th>
        <th style = "text-align: left">مسیر پیدا شده توسط الگوریتم</th>
    </tr>
    <tr>
        <td style = "text-align: left">BFS</td>
        <td style = "text-align: left">66</td>
        <td style = "text-align: left">3976</td>
        <td style = "text-align: left">0.210</td>
        <td style = "text-align: left">RRDDDDDDDDLLRDRRRRRRRRUUUUUUUUULLDDDDDLLDDDDLLLLRRRRUUUUURRRRDDDDD</td>
    </tr>
    <tr>
        <td style = "text-align: left">IDS</td>
        <td style = "text-align: left">66</td>
        <td style = "text-align: left">964088</td>
        <td style = "text-align: left"> 42.44</td>
        <td style = "text-align: left"> RRDDDDDDDDLLRDRRRRRRRRUUUUUUUUULLDDDDDLLDDDDLLLLRRRRUUUUURRRRDDDDD</td>
    </tr>
    <tr>
        <td style = "text-align: left">A*</td>
        <td style = "text-align: left">66</td>
        <td style = "text-align: left">990</td>
        <td style = "text-align: left"> 0.073</td>
        <td style = "text-align: left"> RRDDDDDDLLDDDRRRRUUUURURUUURURRLLDDDDLLDDLDDLDLLRRRRUUUUURRRRDDDDD</td>
    </tr>
        <tr>
        <td style = "text-align: left">A* $\alpha = 2$</td>
        <td style = "text-align: left">66</td>
        <td style = "text-align: left">158</td>
        <td style = "text-align: left"> 0.0085</td>
            <td style = "text-align: left"> RRDDDDDDDDLLRDRRRRRRRRUUUUUUUUULLDDDDDLLDDDDLLLLRRRRUUUUURRRRDDDDD</td>
    </tr>
            <tr>
        <td style = "text-align: left">A* $\alpha = 10$</td>
        <td style = "text-align: left">66</td>
        <td style = "text-align: left">119</td>
        <td style = "text-align: left"> 0.0067</td>
            <td style = "text-align: left"> RRDDDDDDDDLLRDRRRRRRRRUUUUUUUUULLDDDDDLLDDDDLLLLRRRRUUUUURRRRDDDDD</td>
    </tr>
</table>

<html>
    <div dir = "rtl" style="font-size: 22px; color: green">
        <b>
            جمع بندی:
        </b>
    </div>
    <br>
    <br>
    <div dir = "rtl" style = "font-size: 16px">
        <b>
            الگوریتم های جست و جوی زیادی وجود دارند که به دو گروه آگاهانه و ناآگاهانه تقسیم می‌شوند و ما با توجه به نیاز و محدودیت هایمان مانند محدودیت زمانی و حافظه می‌توانیم الگوریتم مناسب را انتخاب و پیاده سازی کنیم.
            <br>
            می‌توان به صورت کلی گفت که الگوریتم های آگاهانه از لحاظ پیچیدگی زمانی و حافظه مصرفی نسبت به الگوریتم های نا آگاهانه کارآمد تر و کم هزینه تر هستند چون از اطلاعات محیط استفاده می‌کنند(در غالب
            Heuristic function)
        </b>
    </div>
</html>

<div dir = "rtl" style = "font-size: 20px; font-family: Calibri; color: green;">
    <b>
    ارائه راهکار برای توسعه و بهبود پروژه:
    </b>
    </div>
    <br>
    </div>
    <div dir = "rtl" style = "font-size: 16px; font-family: Calibri;">
    <b>
       به نظر من بهتر بود به جای محدودیت زمانی محدودیت روی تعداد استیت های دیده شده گذاشته می‌شد چون طریقه ی پیاده سازی خیلی روی زمان اجرا تاثیر گذار است.
    </b>
</div>


<div dir = "rtl" style = "font-size: 20px; color: green">
    <b>
    منابع:
    </b>
    </div>
    </div>
    <div style = "font-size: 16px">
    <b>
      https://stackoverflow.com/
        <br>
        https://www.w3schools.com/
    </b>
</div>